src/libostree/ostree-repo.h

```diff
@@ -808,6 +808,7 @@ typedef enum {
   OSTREE_REPO_CHECKOUT_OVERWRITE_NONE = 0,
   OSTREE_REPO_CHECKOUT_OVERWRITE_UNION_FILES = 1,
   OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES = 2, /* Since: 2017.3 */
+  OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_DIRS = 3 /* Since: 2017.11 */
```

Comment:
This might be a bit bikeshed, though `ADD_DIRS` seems funny to me, since it also adds files in already existing dirs. How about something like `DISJOINT_UNION_FILES` instead? E.g. description could be: `When layering checkouts, error out on exiting files, and do not modify existing directories`.

Reply:
Hey, thanks for the comment :). I am fine with both names, but if I am a bit unsure I am getting the description correctly. Does it mean that we dont want to modify the existing directories at all upon failure? ( we need to clean up the old files checked out upon failures) 
 because ... we current still do not revert changes done to existing directories upon failure for checkout. 

 Since checkout is done recursively, it is likely that before the checkout errors out, some files have been already checked out into the file system. In that case, we kind of still "modify" existing directories.  Here is a small example of the case I mentioned:
```rm -rf ttt/
[vagrant@localhost ~]$ ls
atomic  CLearning  fedora-atomic  GLibTrial  libdnf  ostree  PythonTesting  repo  rpm-ostree  test  yum
[vagrant@localhost ~]$ sudo ostree checkout test_branch ttt
[vagrant@localhost ~]$ ls
atomic  CLearning  fedora-atomic  GLibTrial  libdnf  ostree  PythonTesting  repo  rpm-ostree  test  ttt  yum
[vagrant@localhost ~]$ cd ttt
[vagrant@localhost ttt]$ ls
a  b
[vagrant@localhost ttt]$ rm a
[vagrant@localhost ttt]$ cd
[vagrant@localhost ~]$ sudo ostree checkout --add-dirs test_branch ttt
error: Hardlinking fb/0029f0a483777d541db42de53320e46e67ffd7df21203087766d966509f14b.file to b: File exists
[vagrant@localhost ~]$ ls
atomic  CLearning  fedora-atomic  GLibTrial  libdnf  ostree  PythonTesting  repo  rpm-ostree  test  ttt  yum
[vagrant@localhost ~]$ cd ttt
[vagrant@localhost ttt]$ ls
a  b
```

I did not touch any other error handling part for original code, so I am guessing this problem( the files remaining) was there before? If that is not the case, please let me know, so I can go ahead and fix it, I am also willing to add the cleanup if needed. :)

Reply:
Yeah, the "do not modify existing directories" could be reworded. In this case, it essentially means "don't error out if the directory already exists"; so changing the set of files in the directory itself is fine.

Anyway, re. the new flag name itself, I think it's OK either way. As long as it's properly documented!

Reply:
yup, agreed, will change the comment to make it more understandable :)

Reply:
Right, we don't try at all to handle partial checkout errors; the way libostree programs all tend to work is to make a *new* root and throw the entire thing away if something failed.

Reply:
As far as the name...I spent some time on https://en.wikipedia.org/wiki/Disjoint_union and was confused until I realized it meant the second definition - simply that the sets of files are disjoint and we're unioning them, right?

I suspect either way the doc text is most important.  Maybe:

> When layering checkouts, error out if a file would be replaced, but add new files and directories

?

Reply:
yup sure, sounds good to me, doing it. I was not too sure about disjoint_union earlier too, so thanks for the clarification as well. :)